### PR TITLE
Add resiliency to service invocation

### PR DESCRIPTION
**Pull request by halspang**
# Description

Add resiliency to service invocation.

Signed-off-by: Hal Spang <halspang@microsoft.com>

## Issue reference

Please reference the issue this PR will close: Resiliency

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by yaron2** [comment]
This needs to cover the gRPC proxy path too.

**Comment by yaron2** [comment]
> > This needs to cover the gRPC proxy path too.
> 
> Yup, I was going to include it in a separate PR but I can also add it here since it's just another version of service invocation.

Better to add here

**Comment by halspang** [comment]
@yaron2 - Can you recommend where to put the retries for proxying? I tried in `handler.go`, which seemed to be the natural case. However, when I retried within the `forward_to_` methods, the retry never seemed to trigger a request. When I retried the around the entire handler functionality, I kept getting errors like the following for each retry:

```
== APP == fail: Grpc.AspNetCore.Server.ServerCallHandler[14]                                                                                                                                                                                                
== APP ==       Error reading message.                                                                                                                                                                                                                      
== APP == Grpc.Core.RpcException: Status(StatusCode=Internal, Detail="Incomplete message.")  
```

I thought this may be because I was closing the stream, but I have the stream construction within the retry. Any direction here would be appreciated!

**Comment by yaron2** [comment]
> @yaron2 - Can you recommend where to put the retries for proxying? I tried in `handler.go`, which seemed to be the natural case. However, when I retried within the `forward_to_` methods, the retry never seemed to trigger a request. When I retried the around the entire handler functionality, I kept getting errors like the following for each retry:
> 
> ```
> == APP == fail: Grpc.AspNetCore.Server.ServerCallHandler[14]                                                                                                                                                                                                
> == APP ==       Error reading message.                                                                                                                                                                                                                      
> == APP == Grpc.Core.RpcException: Status(StatusCode=Internal, Detail="Incomplete message.")  
> ```
> 
> I thought this may be because I was closing the stream, but I have the stream construction within the retry. Any direction here would be appreciated!

Can you try redialing the connection?

**Comment by yaron2** [comment]
> @yaron2 - Can you recommend where to put the retries for proxying? I tried in `handler.go`, which seemed to be the natural case. However, when I retried within the `forward_to_` methods, the retry never seemed to trigger a request. When I retried the around the entire handler functionality, I kept getting errors like the following for each retry:
> 
> ```
> == APP == fail: Grpc.AspNetCore.Server.ServerCallHandler[14]                                                                                                                                                                                                
> == APP ==       Error reading message.                                                                                                                                                                                                                      
> == APP == Grpc.Core.RpcException: Status(StatusCode=Internal, Detail="Incomplete message.")  
> ```
> 
> I thought this may be because I was closing the stream, but I have the stream construction within the retry. Any direction here would be appreciated!

I believe this is where you should be retrying: https://github.com/dapr/dapr/blob/c503710ab2d12de04aae0c86f3031cb187cc453e/pkg/grpc/proxy/handler.go#L107.

Instead of returning the error, you should (most likely) literally `goto` here: https://github.com/dapr/dapr/blob/c503710ab2d12de04aae0c86f3031cb187cc453e/pkg/grpc/proxy/handler.go#L63.

**Comment by halspang** [comment]
@yaron2 - I tried your suggestion and confirmed that we're going through the reconnection path. Still getting the same error though. So we can browse through it, I've put the code up here.

File: pkg/grpc/proxy/handler.go [hunk 1]
```diff
@@ -148,13 +169,22 @@ func (s *handler) forwardServerToClient(src grpc.ServerStream, dst grpc.ClientSt
...
+					break
+				}
+			}
+		}
 		for i := 0; ; i++ {
 			if err := src.RecvMsg(f); err != nil {
 				ret <- err // this can be io.EOF which is happy case
 				break
 			}
+			s.bufferedCalls = append(s.bufferedCalls, f)
```
Comment (yaron2) [inline pkg/grpc/proxy/handler.go:186]: 1. Isn't this a memory leak? we keep adding frames but never cleaning them up.
2. Shouldn't a call be added to `bufferedCalls` only if there's an error?

Comment (halspang) [reply pkg/grpc/proxy/handler.go:186]: We do reset `s.bufferedCalls` at the start of the call, but we could also clean it after we're done.

Regardless, this is actually how the grpc middleware does this (albeit with locking): https://github.com/grpc-ecosystem/go-grpc-middleware/blob/560829fc74fcf9a69b7ab01d484f8b8961dc734b/retry/retry.go#L158

Comment (yaron2) [reply pkg/grpc/proxy/handler.go:186]: Lets clean it as soon as we're done, please :)

Comment (halspang) [reply pkg/grpc/proxy/handler.go:187]: Done!

**Comment by artursouza** [comment]
I will approve after test pass.

**Review by artursouza — APPROVED** [review]

File: pkg/grpc/proxy/handler.go [hunk 2]
```diff
@@ -46,70 +53,98 @@ func RegisterService(server *grpc.Server, director StreamDirector, serviceName s
...
-	}
 
-	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
-	// TODO(mwitkow): Add a `forwarded` header to metadata, https://en.wikipedia.org/wiki/X-Forwarded-For.
-	clientStream, err := grpc.NewClientStream(clientCtx, clientStreamDescForProxying, backendConn, fullMethodName, grpc.CallContentSubtype((&codec.Proxy{}).Name()))
-	if err != nil {
-		return err
+	// Fetch the AppId so we can reference it for resiliency.
+	md, _ := metadata.FromIncomingContext(serverStream.Context())
+	v := md.Get(diagnostics.GRPCProxyAppIDKey)
```
Comment (yaron2) [inline pkg/grpc/proxy/handler.go:85]: Can we move that to the section where a resiliency policy is used?

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: pkg/grpc/proxy/handler.go (hunk 2 above)
Comment (halspang) [reply pkg/grpc/proxy/handler.go:85]: We need it to get the resiliency policy, so it has to be before the policy. Or am I misunderstanding what you're asking?

Comment (yaron2) [reply pkg/grpc/proxy/handler.go:85]: Yeah I realize it's needed for the policy.

**Pull request merged**